lsp/nls/src/field_walker.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use lsp_types::{CompletionItem, CompletionItemKind, Documentation, MarkupContent, MarkupKind};
+use nickel_lang_core::{
+    identifier::Ident,
+    term::{record::FieldMetadata, BinaryOp, RichTerm, Term},
+};
+
+use crate::{linearization::completed::Completed, server::Server};
+
+/// The position at which something is defined.
+#[derive(Clone, Debug)]
+pub struct Def {
+    /// The identifier at the definition site.
+    ///
+    /// Remember that an `Ident` has a position; this one points to the identifier
+    /// at the position where it is bound. For example, in `{ foo = 1 }`, this ident
+    /// might point at the `foo`.
+    pub ident: Ident,
+    /// The value assigned by the definition, if there is one.
+    ///
+    /// For example, in `{ foo = 1 }`, this could point at the `1`.
+    pub value: Option<RichTerm>,
+    /// Field metadata.
+    pub metadata: Option<FieldMetadata>,
+}
+
+impl Def {
+    fn doc(&self) -> Option<Documentation> {
+        let doc = self.metadata.as_ref()?.doc.as_ref()?;
+        Some(Documentation::MarkupContent(MarkupContent {
+            kind: MarkupKind::Markdown,
+            value: doc.clone(),
+        }))
+    }
+
+    // If the field is annotated, returns its type annotation (preferred) or its
+    // contract annotation (fallback).
+    fn detail(&self) -> Option<String> {
+        self.metadata
+            .as_ref()
+            .and_then(|FieldMetadata { annotation, .. }| {
+                annotation
+                    .typ
+                    .as_ref()
+                    .map(|ty| ty.typ.to_string())
+                    .or_else(|| annotation.contracts_to_string())
+            })
+    }
+
+    /// Creates a completion item from this definition.
+    pub fn to_completion_item(&self) -> CompletionItem {
+        /// Attach quotes to a non-ASCII string
+        fn adjust_name(name: &str) -> String {
+            if name.is_ascii() {
+                String::from(name)
+            } else {
+                format!("\"{name}\"")
+            }
```

Comment:
Just `is_ascii` might not be good enough. For example, a field name might contain spaces and should then be quoted in Nickel. Or is this something that the LSP spec takes care of already?

We have `QUOTING_REGEX` in `core/src/pretty.rs` and `KEYWORDS` in `core/src/parser/lexer.rs` to check if an identifier can be used unquoted. Not sure if you can use those without trouble in the LSP, but if not we should probably add some function to `Ident` to help.